### PR TITLE
UK Bank Holidays

### PR DESCRIPTION
**Pull request by bendiggle**
- Add functionality that gets the list of bank holidays and check if the date passed is a bank holiday or not.
- Add unit tests

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+    if (!Array.isArray(countries)) {
+        throw new Error('Countries must be an array');
+    }
+
+    if (!validCountry(countries)) {
+        throw new Error('Country not on the list');
+    }
+
+    date = formatDate(date);
```
Comment (paulgain) [inline src/index.js:30]: I think you need to perform a check to ensure the date string is in the correct format `DD-MM-YYYY`. What happens if I miss off a `Y`?

Comment (bendiggle) [reply src/index.js:30]: @paulgain point. I can do `moment().isValid()` to check

File: src/index.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+    try {
+        const bankHolidays = await request.get('https://www.gov.uk/bank-holidays.json');
```
Comment (paulgain) [inline src/index.js:33]: Going forward we should think about cacheing this call, maybe for 24hrs.

File: src/index.js (hunk 1 above)
Comment (bendiggle) [reply src/index.js:30]: or a regex

Comment (paulgain) [reply src/index.js:30]: Whatever you decide ensure the user has a good error message.

File: src/utils.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    return flatten(dateList);
+};
+
+const validCountry = countries => {
+
+    let valid = true;
+
+    valid = countries.every(country => {
+       return includes(ukCountries, country);
+    });
```
Comment (paulgain) [inline src/utils.js]: Can this be a one-liner where you drop the curly braces and the `return` statement.

File: test/unit/ukBankHolidays.test.js [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+
+        it('should return false when date is as string and not a bank holiday', async function() {
+            const isBankHoliday = await ukBankHolidays(moment('21-03-2018', 'DD-MM-YYYY'), countries);
+            expect(isBankHoliday).to.equal(false);
+        });
+
+    });
+
+    describe('errors', () => {
+
```
Comment (paulgain) [inline test/unit/ukBankHolidays.test.js:202]: As discussed earlier we require a test for when a user accidentally passes in a date string in the wrong format.

**Comment by paulgain** [comment]
Just as a general note we need to think about passing an array of dates and what would the response look like.

File: src/utils.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+};
+
+const isDateInList = (dateList, date) => {
+    return includes(dateList, date);
+};
+
+const bankHolidayList = (dates, countries) => {
+    let dateList = [];
+
+    countries.forEach(country => {
```
Comment (paulgain) [inline src/utils.js]: This should be a map.

File: src/index.js (hunk 2 above)
Comment (bendiggle) [reply src/index.js:33]: Something like this? https://www.npmjs.com/package/memory-cache

**Comment by bendiggle** [comment]
@paulgain yeah true (array of dates). We'd want to return the date that is a bank holiday

File: src/index.js (hunk 2 above)
Comment (paulgain) [reply src/index.js:33]: This appears to be more popular: https://www.npmjs.com/package/node-cache

**Review by paulgain — APPROVED** [review]

**Pull request merged**